book-content/chapters/05-unions-literals-and-narrowing.md

```diff
@@ -0,0 +1,1856 @@
+  const isUsernameOK = typeof username === "string";
+
+  if (isUsernameOK) {
+    return username.length > 5;
+  }
+
+  return false;
+}
+```
+
+TypeScript is smart enough to understand that the value of `isUsernameOK` corresponds to whether `username` is a string or not. Very smart.
+
+All of the above options use `if` statements to perform checks by narrowing types by using `typeof`.
+
+No matter which option you go with, remember that you can always use an `if` statement to narrow your type and add code to the case that the condition passes.
+
+#### Solution 2: Throwing Errors to Narrow
+
+The issue with this code is that `document.getElementById` returns `null | HTMLElement`. But we want to make sure that `appElement` is an `HTMLElement` before we use it.
+
+We are pretty sure that `appElement` exists. If it doesn't exist, we probably want to crash the app early so that we can get an informative error about what's gone wrong.
+
+So, we can add an `if` statement that checks if `appElement` is falsy, then throws an error:
+
+```typescript
+if (!appElement) {
+  throw new Error("Could not find app element");
+}
+```
+
+By adding this error condition, we can be sure that we will never reach any subsequent code if `appElement` is `null`.
+
+If we hover over `appElement` after the `if` statement, we can see that TypeScript now knows that `appElement` is an `HTMLElement` - it's no longer `null`. This means our test also now passes:
+
+```ts twoslash
+import { Equal, Expect } from "@total-typescript/helpers";
+
+const appElement = document.getElementById("app");
+
+if (!appElement) {
+  throw new Error("Could not find app element");
+}
+
+// ---cut---
+console.log(appElement);
+//          ^?
+
+type Test = Expect<Equal<typeof appElement, HTMLElement>>; // passes
+```
+
+Throwing errors like this can help you identify issues at runtime. In this specific case, it narrows down the code _outside_ of the immediate `if` statement scope. Amazing.
+
+#### Solution 3: Using `in` to Narrow
+
+Your first instinct will be to check if `response.data` is truthy.
+
+```ts twoslash
+// @errors: 2339
+type APIResponse =
+  | {
+      data: {
+        id: string;
+      };
+    }
+  | {
+      error: string;
+    };
+
+// ---cut---
+const handleResponse = (response: APIResponse) => {
+  if (response.data) {
+    return response.data.id;
+  } else {
+    throw new Error(response.error);
+  }
+};
+```
+
+But you'll get an error. This is because `response.data` is only available on one of the members of the union. TypeScript doesn't know that `response` is the one with `data` on it.
+
+##### Option 1: Changing the Type
+
+It may be tempting to change the `APIResponse` type to add `.data` to both branches:
+
+```tsx
+type APIResponse =
+  | {
+      data: {
+        id: string;
+      };
+    }
+  | {
+      data?: undefined;
+      error: string;
+    };
+```
+
+This is certainly one way to handle it. But there is a built-in way to do it.
```

Comment:
+ to make reverse checks possible and all you should add `error?: undefined` to the other member